### PR TITLE
*: add foreign key constraint check when execute insert statement

### LINKED ISSUE #37465: support foreign key constrain check when execute DML
# Dev Task

- [ ] Add foreign key constraint check when insert/update/delete into foreign key child/parent table.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #37465

### What is changed and how it works?

Add foreign key constraint check when insert/update/delete into foreign key child/parent table.

More design detail see the design doc:  https://github.com/pingcap/tidb/pull/35982

Related Integration Test: https://github.com/pingcap/tidb-test/pull/1965

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/foreign_key.go [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/set"
+)
+
+// WithForeignKeyTrigger indicates the executor has foreign key check or cascade.
+type WithForeignKeyTrigger interface {
+	GetFKChecks() []*FKCheckExec
```
Comment (you06) [inline executor/foreign_key.go:35]: Only `InsertExec` implements this interface, so other statements like `update` won't check foreign keys, is it ok?

Comment (morgo) [reply executor/foreign_key.go:35]: That looks to be the case. Here is the testcase from the [MySQL manual](https://dev.mysql.com/doc/mysql-tutorial-excerpt/8.0/en/example-foreign-keys.html):

```sql
DROP TABLE IF EXISTS parent, child;
SET foreign_key_checks = 1;
SET GLOBAL tidb_enable_foreign_key=1;
CREATE TABLE parent (
    id INT NOT NULL,
    PRIMARY KEY (id)
) ENGINE=INNODB;
CREATE TABLE child (
    id INT,
    parent_id INT,
    INDEX par_ind (parent_id),
    FOREIGN KEY (parent_id)
        REFERENCES parent(id)
) ENGINE=INNODB;
INSERT INTO parent (id) VALUES (1);
INSERT INTO child (id,parent_id) VALUES (1,1);
INSERT INTO child (id,parent_id) VALUES(2,2); # error expected (works)
DELETE FROM parent WHERE id = 1; # error expected (not supported yet)
```

May I suggest updating the PR title to just be for `INSERT` and not DML?

Comment (crazycs520) [reply executor/foreign_key.go:35]: Great, currently this PR only supports foreign key check in `INSERT`.
Before the commit [403aaa89fde08f2238a80dca4eb44bd6e98b916b](https://github.com/pingcap/tidb/pull/37466/commits/403aaa89fde08f2238a80dca4eb44bd6e98b916b), this PR was support foreign key check in `INSERT/UPDATE/DELETE`, but the PR is so large, then I only support foreign key check in `INSERT`.

File: executor/foreign_key.go [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+	}
+	sessVars := fkc.ctx.GetSessionVars()
+	lockCtx, err := newLockCtx(fkc.ctx, sessVars.LockWaitTimeout, len(fkc.toBeLockedKeys))
+	if err != nil {
+		return err
+	}
+	// WARN: Since tidb current doesn't support `LOCK IN SHARE MODE`, therefore, performance will be very poor in concurrency cases.
+	// TODO(crazycs520):After TiDB support `LOCK IN SHARE MODE`, use `LOCK IN SHARE MODE` here.
+	forUpdate := atomic.LoadUint32(&sessVars.TxnCtx.ForUpdate)
+	err = doLockKeys(ctx, fkc.ctx, lockCtx, fkc.toBeLockedKeys...)
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/foreign_key.go:143]: Yes, the behavior is just like `SELECT FOR SHARE`, both optimistic and pessimistic transactions are needs to lock the keys.

But TiDB currently does not support `LOCK IN SHARE MORE`, only supports `LOCK FOR UPDATE MODE`. So the performance will be bad when there is some lock contention.

File: executor/foreign_key.go [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+	// 	+----+--------+--------+
+	for _, val := range vals {
+		if val.IsNull() {
+			return true
+		}
+	}
+	return false
+}
+
+func getColumnsOffsets(tbInfo *model.TableInfo, cols []model.CIStr) ([]int, error) {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/foreign_key.go]: great, done.

**Review by you06 — COMMENTED** [review]
The only concern for me is the pessimistic locks in optimistic transactions, can we check it in the prewrite phase instead?
Rest LGTM

File: store/driver/txn/snapshot.go [hunk 4]
```diff
@@ -130,6 +130,11 @@ func (s *tikvSnapshot) SetOption(opt int, val interface{}) {
 		s.KVSnapshot.SetRequestSourceType(val.(string))
 	case kv.ReplicaReadAdjuster:
 		s.KVSnapshot.SetReplicaReadAdjuster(val.(txnkv.ReplicaReadAdjuster))
+	case kv.ScanBatchSize:
+		size := val.(int)
+		if size > 0 {
+			s.KVSnapshot.SetScanBatchSize(size)
```
Comment (AilinKid) [inline store/driver/txn/snapshot.go:136]: what‘s the use of this size

File: executor/foreign_key_test.go [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@global.tidb_enable_foreign_key=1")
+	tk.MustExec("set @@foreign_key_checks=1")
+	tk.MustExec("use test")
+
+	tk.MustExec("CREATE TABLE t1 (i INT PRIMARY KEY);")
+	tk.MustExec("CREATE TABLE t2 (i INT, FOREIGN KEY (i) REFERENCES t1 (i));")
+	tk.MustExec("INSERT INTO t1 VALUES (1),(3);")
+	tk.MustExec("INSERT IGNORE INTO t2 VALUES (1),(2),(3),(4);")
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/foreign_key_test.go:380]: In my local test with MySQL 8.0, mysql has wanings here:

```sql
mysql> INSERT IGNORE INTO t2 VALUES (1),(2),(3),(4);
Query OK, 2 rows affected, 2 warnings (0.00 sec)
Records: 4  Duplicates: 2  Warnings: 2

mysql> show warnings;
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                        |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`i`) REFERENCES `t1` (`i`)) |
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`i`) REFERENCES `t1` (`i`)) |
+---------+------+------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.26    |
+-----------+
1 row in set (0.00 sec)
```

File: planner/core/foreign_key.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	}
+	if tblIdx == nil {
+		return nil, failedErr
+	}
+
+	var handleCols []*table.Column
+	if referTbIdxInfo.Primary && tblInfo.IsCommonHandle {
+		cols := tbl.Cols()
+		for _, idxCol := range referTbIdxInfo.Columns {
+			handleCols = append(handleCols, cols[idxCol.Offset])
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/foreign_key.go]: Nice catch, It is indeed useless and been deleted now.

File: store/driver/txn/snapshot.go (hunk 4 above)
Comment (crazycs520) [reply store/driver/txn/snapshot.go:136]: `KVSnapshot.Iter` use `Scanner` to scan key/values from storage(TiKV), the `ScanBatchSize` uses to control to fetch how many kv in one RPC request.

File: executor/insert_common.go [hunk 7]
```diff
@@ -1126,6 +1128,13 @@ func (e *InsertValues) batchCheckAndInsert(ctx context.Context, rows [][]types.D
 			defer snapshot.SetOption(kv.CollectRuntimeStats, nil)
 		}
 	}
+	sc := e.ctx.GetSessionVars().StmtCtx
+	for _, fkc := range e.fkChecks {
```
Comment (AilinKid) [inline executor/insert_common.go:1132]: any test case about insert into select from and load data？

File: executor/foreign_key.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+
+func (fkc *FKCheckExec) checkIndexKeys(ctx context.Context, txn kv.Transaction) error {
+	if len(fkc.toBeCheckedPrefixKeys) == 0 {
+		return nil
+	}
+	memBuffer := txn.GetMemBuffer()
+	snap := txn.GetSnapshot()
+	snap.SetOption(kv.ScanBatchSize, 2)
+	defer func() {
+		snap.SetOption(kv.ScanBatchSize, 256)
```
Comment (AilinKid) [inline executor/foreign_key.go:247]: I see the `const defaultScanBatchSize = 128`

**Comment by AilinKid** [comment]
Rest LGTM

File: executor/foreign_key.go (hunk 8 above)
Comment (crazycs520) [reply executor/foreign_key.go:247]: after https://github.com/tikv/client-go/pull/588 merged, I will use the `DefaultScanBatchSize` instead `256` here.

File: executor/insert_common.go (hunk 7 above)
Comment (crazycs520) [reply executor/insert_common.go:1132]: Already add test for `insert  into select`.

Currently `load data` doesn't do foreign check, so can not add test for `load data`.

**Review by AilinKid — APPROVED** [review]

**Comment by AilinKid** [comment]
@morgo take another look again

File: executor/foreign_key.go (hunk 8 above)
Comment (you06) [reply executor/foreign_key.go:247]: Can we use `DefaultScanBatchSize` instead of a magic number?

Comment (crazycs520) [reply executor/foreign_key.go]: done.

**Review by you06 — APPROVED** [review]

**Review by morgo — APPROVED** [review]

**Pull request merged**